cli/src/cmd/forge/script/receipts.rs

```diff
@@ -2,83 +2,118 @@ use crate::{
     cmd::forge::script::sequence::ScriptSequence, init_progress, update_progress,
     utils::print_receipt,
 };
-use ethers::prelude::{PendingTransaction, TxHash};
+use ethers::{
+    prelude::{PendingTransaction, TxHash},
+    providers::Middleware,
+    types::TransactionReceipt,
+};
 use foundry_common::RetryProvider;
 use futures::StreamExt;
 use std::sync::Arc;
-use tracing::trace;
+use tracing::{trace, warn};
+
+/// Convenince enum for internal signalling of transaction status
+enum TxStatus {
+    Dropped,
+    Success(TransactionReceipt),
+    Revert(TransactionReceipt),
+}
+
+impl From<TransactionReceipt> for TxStatus {
+    fn from(receipt: TransactionReceipt) -> Self {
+        let status = receipt.status.expect("receipt is from an ancient, pre-EIP658 block");
+        if status.is_zero() {
+            TxStatus::Revert(receipt)
+        } else {
+            TxStatus::Success(receipt)
+        }
+    }
+}
 
-/// Gets the receipts of previously pending transactions.
+/// Gets the receipts of previously pending transactions, or removes them fro
+/// the deploy sequence's pending vector
 pub async fn wait_for_pending(
     provider: Arc<RetryProvider>,
     deployment_sequence: &mut ScriptSequence,
 ) -> eyre::Result<()> {
-    if !deployment_sequence.pending.is_empty() {
-        println!("##\nChecking previously pending transactions.");
-        wait_for_receipts(deployment_sequence.pending.clone(), deployment_sequence, provider)
-            .await?;
+    if deployment_sequence.pending.is_empty() {
+        return Ok(())
     }
-    Ok(())
+    println!("##\nChecking previously pending transactions.");
+    clear_pendings(provider, deployment_sequence, None).await
 }
 
-/// Waits for a list of receipts. If it fails, it tries to retrieve the transaction hash that can be
-/// used on a later run with `--resume`.
-pub async fn wait_for_receipts(
-    tx_hashes: Vec<TxHash>,
-    deployment_sequence: &mut ScriptSequence,
+/// Traverses a set of pendings and either finds receipts, or clears them from
+/// the deployment sequnce.
+///
+/// If no `tx_hashes` are provided, then `deployment_sequence.pending` will be
+/// used. For each `tx_hash`, we check if it has confirmed. If it has
+/// confirmed, we push the receipt (if successful) or push an error (if
+/// revert). If the transaction has not confirmed, but can be found in the
+/// node's mempool, we wait for its receipt to be available. If the transaction
+/// has not confirmed, and cannot be found in the mempool, we remove it from
+/// the `deploy_sequence.pending` vector so that it will be rebroadcast in
+/// later steps.
+pub async fn clear_pendings(
     provider: Arc<RetryProvider>,
+    deployment_sequence: &mut ScriptSequence,
+    tx_hashes: Option<Vec<TxHash>>,
```

Comment:
I was looking at the main branch code and this `tx_hashes` arg really doesn't seem necessary.  Any `TxHash` here should be in `deployment_sequence.pending` even when calling it with `--slow/sequential_broadcast`.

Reply:
there is one spot that used it for something than the pending vec from the sequence, but i didn't look into what exactly it was doing